grpc-server/hardware.go

```diff
@@ -297,22 +309,35 @@ func (s *server) Delete(ctx context.Context, in *hardware.DeleteRequest) (*hardw
 }
 
 func (s *server) validateHardwareData(ctx context.Context, hw *hardware.Hardware) error {
-	interfaces := hw.GetNetwork().GetInterfaces()
-	for i := range hw.GetNetwork().GetInterfaces() {
-		data, _ := s.db.GetByMAC(ctx, interfaces[i].GetDhcp().GetMac())
-		if data != "" {
-			logger.With("MAC", interfaces[i].GetDhcp().GetMac()).Info(duplicateMAC)
+	for _, iface := range hw.GetNetwork().GetInterfaces() {
+		mac := iface.GetDhcp().GetMac()
+
+		if data, _ := s.db.GetByMAC(ctx, mac); data != "" {
+			logger.With("MAC", mac).Info(duplicateMAC)
+
 			newhw := hardware.Hardware{}
-			err := json.Unmarshal([]byte(data), &newhw)
-			if err != nil {
+			if err := json.Unmarshal([]byte(data), &newhw); err != nil {
 				logger.Error(err, "Failed to unmarshal hardware data")
 				return err
 			}
+
 			if newhw.Id == hw.Id {
 				return nil
 			}
-			return errors.New(fmt.Sprintf(conflictMACAddr, interfaces[i].GetDhcp().GetMac()))
+
+			return errors.New(fmt.Sprintf(conflictMACAddr, mac))
 		}
 	}
+
 	return nil
 }
+
+func normalizeHardwareData(hw *hardware.Hardware) {
+	// Ensure MAC is stored as lowercase
+	for _, iface := range hw.GetNetwork().GetInterfaces() {
```

Comment:
I switched things to use the generated proto getters rather than nil checks, since they already include the nil checks and it also eliminates the use of slice indexing.